docs/design-documents/features/storage/SecureStore/SecureStore_design.md

```diff
@@ -0,0 +1,448 @@
+typedef struct {
+    KVStore::iterator_t underlying_it;
+} key_iterator_handle_t;
+```
+
+
+### Initialization and reset
+
+**init function**
+
+Header:  
+`virtual int init();`
+
+Pseudo code:  
+- if `_is_initialized` return OK
+- Take `_mutex`
+- Initialize `_entropy` with TLS entropy APIs
+- Using `DeviceKey` APIs, get the device key
+- Allocate `_scratch_buf` as a 32 byte array
+- Set `_is_initialized` to true
+- Release `_mutex`     
+  
+**deinit function**
+
+Header:  
+`virtual int deinit();`
+
+Pseudo code:  
+- if not `_is_initialized` return OK
+- Take `_mutex` 
+- Deinitialize `_entropy`
+- Deallocate `_encrypt_key`, `_auth_key`, `_scratch_buf`
+- Release `_mutex`   
+
+**reset function**
+
+Header:  
+`virtual int reset();`
+
+Pseudo code:  
+- Take `_mutex` 
+- Call `_underlying_kv` `reset` API  
+- Call `_rbp_kv` `reset` API  
+- Release `_mutex`   
+
+### Core APIs
+
+**set function**
+
+Header:  
+`virtual int set(const char *key, const void *buffer, size_t size, uint32_t create_flags);`
+
+Pseudo code:  
+- Call `set_start` with all fields and a local `set_handle_t` variable
+- Call `set_add_data` with `buffer` and `size`
+- Call `set_finalize`
+- Return OK
+
+**get function**
+
+Header:  
+`virtual int get(const char *key, void *buffer, size_t buffer_size, size_t *actual_size = NULL, size_t offset = 0);`
+
+Pseudo code:  
+- if not `_is_initialized` return error
+- Take `_mutex`
+- Call `_underlying_kv` `get` API with `metadata` size into a `metadata` local structure
+- If failure
+	- If rollback protection flag set
+		- Call `_rbp_kv` `get` API on a local `rbp_cmac` variable, key is `key`, size 16
+		- If no error, return "RBP authentication" error
+	- Return "Key not found error"
+- If authentication flag set
+	- Derive a key from device key and `key`
+	- Allocate and initialize `auth_handle` CMAC calculation local handle with derived key 
+	- Using `auth_handle` handle, calculate CMAC on `key` & `metadata`
+- If encrypt flag set
+	- Derive a key from device key and `key`
+	- Allocate and initialize a local `enc_handle` AES-CTR local handle with derived key and `nonce` field 
+- Set `data_size` local variable to data size in metadata
+- Set `actual_size` to the minimum of `buffer_size` and `data_size`
+- Set `current_offset` to 0
+- While `data_size` > 0
+	- If `current_offset` between `offset` and `actual_size`
+		- Set `dest_buf` to `buffer` and `chunk_size` to `actual_size`
+	- Else
+		- Set `dest_buf` to `_scratch_buf` and `chunk_size` to `actual_size`
+	- Call `_underlying_kv` `get` API with `dest_buf` and `chunk_size`
+	- If authentication flag set, calculate CMAC on `dest_buf`, using `_auth_handle` handle
+	- If encrypt flag set, decrypt `dest_buf` (in place)  using `_enc_handle` handle
+	- Decrement `data_size` by `chunk_size`
+- Call `_underlying_kv` `get` API with on a local `read_cmac` variable, size 16
+- Generate CMAC on local `cmac` variable 
+- If `read_cmac` doesn't match `cmac`, return "data corrupt error"
```

Comment:
This compare needs to be a constant-time compare. This is harder than it sounds

Reply:
Please elaborate. Not sure I understand.

Reply:
a naive compare would look something like this **NOTE: THIS IS BAD. DO NOT USE**:
```C
int pass = 1;
for (int i = 0; i < mac_len; i++) {
    if(mac[i] != reference_mac[i]) {
        pass = 0;
        break;
    }
}
return pass;
```

However, attacker could use this to measure time differentials in different attempts, and then glean information about the underlying key.
A better (but still not sufficient) implementation **NOTE: THIS IS STILL BAD DO NOT USE**:
```C
int pass = 1;
for (int i = 0; i < mac_len; i++) {
    if(mac[i] != reference_mac[i]) {
        pass = 0;
    }
}
return pass;
```
Notice that now it **should** take the same amount of time regardless of where the first miscompare is. However, some compilers may optimize this and create a problem.
MbedTLS has a constant time comparison function `mbedtls_ssl_safer_memcmp()` which should be used for this.

Reply:
Thanks for the clarification. Will use this function.